src/layering.jl

```diff
@@ -1,9 +1,36 @@
 # This contains algorithms for breaking up a DAG into layers
-
-function layer_by_longest_path_to_source(graph)
+"Calculate the layer of each node"
+function layer_by_longest_path_to_source(graph, opt_layer_assign)
     dists = longest_paths(graph, sources(graph))
-    layer_groups = IterTools.groupby(i->dists[i], sort(vertices(graph), by=i->dists[i]))
-    return collect.(layer_groups)
+    layer_groups = collect.(IterTools.groupby(i->dists[i], sort(vertices(graph), by=i->dists[i])))
+    agree_with_opt_layer_assign!(layer_groups, graph, opt_layer_assign)
+    return layer_groups
+end
+
+"Correct the layer of each node, according to the optional parameter by user"
+function agree_with_opt_layer_assign!(layer_groups, graph, opt_layer_assign)
+    if length(opt_layer_assign) > 0
+        key_opt = collect(keys(opt_layer_assign))
+        values_opt = collect(values(opt_layer_assign))
+
+        opt_ids_by_layers = sortperm(collect(values(opt_layer_assign)), rev=true)
+        key_opt = key_opt[opt_ids_by_layers]
+        values_opt = values_opt[opt_ids_by_layers]
+
+        # change layers if needed
+        for (k, l) in zip(key_opt, values_opt)
+            #current layer of node
+            curr_layer = findfirst(k .∈ layer_groups)
+            if curr_layer >= l
+                @warn "Ignored opt_layer_assign for node $k; dists[k] ($(dists[k])) > l ($l)"
```

Comment:
In the case the user specify a layer lower than the current one, the optional value is ignored with a warning